pkg/config/apiserver.go

```diff
@@ -0,0 +1,212 @@
+// The server uses KVStore to persists keys.
+
+import (
+	"fmt"
+	"io"
+	"io/ioutil"
+	"net/http"
+	"strconv"
+
+	restful "github.com/emicklei/go-restful"
+	"github.com/golang/glog"
+	rpc "github.com/googleapis/googleapis/google/rpc"
+
+	"istio.io/mixer/pkg/adapter"
+	"istio.io/mixer/pkg/config/descriptor"
+	pb "istio.io/mixer/pkg/config/proto"
+	"istio.io/mixer/pkg/expr"
+	"istio.io/mixer/pkg/status"
+)
+
+type validateFunc func(cfg map[string]string) (rt *Validated, desc descriptor.Finder, ce *adapter.ConfigErrors)
+
+type readBodyFunc func(r io.Reader) ([]byte, error)
+
+// API is the server wrapper that listens for incoming requests to the manager and processes them
+type API struct {
+	version  string
+	rootPath string
+
+	// used at the back end for validation and storage
+	store    KVStore
+	validate validateFunc
+
+	// house keeping
+	handler http.Handler
+	Server  *http.Server
+
+	// fault injection
+	readBody readBodyFunc
+}
+
+// DefaultAPIPort default port exposed by the API server.
+const DefaultAPIPort = 9094
+
+// register routes
+func (a *API) register(c *restful.Container) {
+	ws := &restful.WebService{}
+	ws.Consumes(restful.MIME_JSON, "application/yaml", "application/x-yaml")
+	ws.Produces(restful.MIME_JSON)
+	ws.Path(a.rootPath)
+
+	ws.Route(ws.
+		GET("/scopes/{scope}/subjects/{subject}/rules").
+		To(a.getRules).
+		Doc("Gets rules associated with the given scope and subject").
+		Param(ws.PathParameter("scope", "scope").DataType("string")).
+		Param(ws.PathParameter("subject", "subject").DataType("string")).
+		Writes(pb.ServiceConfig{}))
+
+	ws.Route(ws.
+		PUT("/scopes/{scope}/subjects/{subject}/rules").
+		To(a.putRules).
+		Doc("Replaces rules associated with the given scope and subject").
+		Param(ws.PathParameter("scope", "scope").DataType("string")).
+		Param(ws.PathParameter("subject", "subject").DataType("string")).
+		Writes(rpc.Status{}))
+	c.Add(ws)
+}
+
+// NewAPI creates a new API server
+func NewAPI(version string, port int, eval expr.Validator, aspectFinder AspectValidatorFinder,
+	builderFinder BuilderValidatorFinder, findAspects AdapterToAspectMapper, store KVStore) *API {
+	c := restful.NewContainer()
+	a := &API{
+		version:  version,
+		rootPath: fmt.Sprintf("/api/%s", version),
+		store:    store,
+		readBody: ioutil.ReadAll,
+		validate: func(cfg map[string]string) (*Validated, descriptor.Finder, *adapter.ConfigErrors) {
+			v := newValidator(aspectFinder, builderFinder, findAspects, true, eval)
+			rt, ce := v.validate(cfg)
+			return rt, v.descriptorFinder, ce
+		},
+	}
+	a.register(c)
+	a.Server = &http.Server{Addr: ":" + strconv.Itoa(port), Handler: c}
+	a.handler = c
+	return a
+}
+
+// Run calls listen and serve on the API server
+func (a *API) Run() {
+	glog.Infof("Starting Config API Server at %v", a.Server.Addr)
+	glog.Warning(a.Server.ListenAndServe())
+}
+
+// getRules returns the entire service config document for the scope and subject
```

Comment:
I think we need to find a different word than "service config" here, since the selector model is more general than that.

So we have "global config" to describe the mixer's overall configuration, and then perhaps "filtered config" to refer to the config state post-selector.